libtest/template_lib.c

```diff
@@ -146,49 +146,63 @@ compile_template(const gchar *template, gboolean escaping)
 }
 
 void
-assert_template_format(const gchar *template, const gchar *expected)
+assert_template_format_with_escaping_and_context_msgs(const gchar *template, gboolean escaping,
+                                                      const gchar *expected, gssize expected_len,
+                                                      LogMessage **msgs, gint num_messages)
 {
-  assert_template_format_with_escaping(template, FALSE, expected);
+  LogTemplate *templ = compile_template(template, escaping);
+  if (!templ)
+    return;
+
+  GString *res = g_string_sized_new(128);
+  const gchar *context_id = "test-context-id";
+
+  log_template_format_with_context(templ, msgs, num_messages, NULL, LTZ_LOCAL, 999, context_id, res);
+  expect_nstring(res->str, res->len,
+                 expected, expected_len >= 0 ? expected_len : strlen(expected),
+                 "context template test failed, template=%s", template);
+  log_template_unref(templ);
+  g_string_free(res, TRUE);
 }
 
 void
-assert_template_format_msg(const gchar *template, const gchar *expected,
-                           LogMessage *msg)
+assert_template_format_with_context_msgs(const gchar *template, const gchar *expected, LogMessage **msgs,
+                                         gint num_messages)
 {
-  assert_template_format_with_escaping_msg(template, FALSE, expected, msg);
+  assert_template_format_with_escaping_and_context_msgs(template, FALSE, expected, -1, msgs, num_messages);
 }
 
 
 void
 assert_template_format_with_escaping_msg(const gchar *template, gboolean escaping,
-                                         const gchar *expected, LogMessage *msg)
+                                         const gchar *expected,
+                                         LogMessage *msg)
 {
-  LogTemplate *templ = compile_template(template, escaping);
-  if (!templ)
-    return;
-
-  GString *res = g_string_sized_new(128);
-  const gchar *context_id = "test-context-id";
-
-  log_template_format(templ, msg, NULL, LTZ_LOCAL, 999, context_id, res);
-  expect_nstring(res->str, res->len, expected, strlen(expected),
-                 "template test failed, template=%s", template);
-  log_template_unref(templ);
-  g_string_free(res, TRUE);
+  assert_template_format_with_escaping_and_context_msgs(template, escaping, expected, -1, &msg, 1);
 }
 
 void
-assert_template_format_with_escaping(const gchar *template, gboolean escaping,
-                                     const gchar *expected)
+assert_template_format_with_escaping(const gchar *template, gboolean escaping, const gchar *expected)
 {
   LogMessage *msg = create_sample_message();
 
   assert_template_format_with_escaping_msg(template, escaping, expected, msg);
```

Comment:
It would worth set log_template_set_escape(templ, escaping) here instead of compile_template instead of passing around. It would reduce the argument list of many functions. It is a bit hard to follow these functions due to the long name and parameter list.

Reply:
I agree. The entire libtest should be given some more thought. It is coming from a pre-criterion time, all I did this with this patch was to rebase it from an ancient pull request #531.

I would prefer not to do this refactor at this point, rather to do it as separate pull request in the future. (famous words I know, but I wouldn't want to block this PR because of it. It has been in purgatory for all eternity).

Reply:
Ok, fine for me.